### PR TITLE
Layla - Edges - Hotel 

### PR DESCRIPTION
**Pull request by lbristol88**
# Hotel
Congratulations! You're submitting your assignment!

## Comprehension Questions
Question | Answer
:------------- | :-------------
Describe a design decision you had to make when working on this project. What options were you considering? What helped you make your final decision? | A design decision I had to make while working on this project was determining what classes would be best for my booking system. I initially was going to create a room class but it didn't really make sense the more I worked on the project. Therefore my finally classes were reservation, daterange, block and the bookingsystem.
Describe a concept that you gained more clarity on as you worked on this assignment. | Determining what kind of classes to make and ensuring that my classes possessed single responsibility. It was good practice to make sure I did not use too many methods in one class. 
Describe a _nominal test_ that you wrote for this assignment. | To start out with creating my tests, I created basic ones that made sure the test would recognize the attributes of each class. For example, I made tests in reservation that made sure a new reservation object contained a room number, price, etc. 
Describe an _edge case test_ that you wrote for this assignment. | I created an edge case test in the booking system to not allow a room number that was outside the available range. For example, an administrator couldn't book a room in the hotel with a room number of 50, since the max room number was 20. If the administrator tried to, it would raise an argument error. 
How do you feel you did in writing pseudocode first, then writing the tests and then the code? | Pseudocode was helpful with designing the big picture of all the components that go into making a reservation and what information a booking system would need in order to function. I wrote a lot of basic tests that were empty at first just to make sure they would pass and ensure I didn't have typos or syntax errors for when I go in to add actual testing data. This was helpful to make sure I don't forget any "do"s or "end"s etc.

### CODE REVIEW

File: lib/block.rb [hunk 1]
```diff
@@ -0,0 +1,18 @@
+module Hotel
+  class BlockOfRooms
+    attr_reader :room_numbers, :date_range, :room_rate
+
+    MAX_ROOM_BLOCK = 5
+
+    def initilize(room_rate, date_range, room_numbers)
+      unless room_numbers.length <= MAX_ROOM_BLOCK
+        raise ArgumentError
+      end
```
Comment (droberts-sea) [inline lib/block.rb:10]: You should always include a message when you raise an error. That way whoever messed up and triggered it knows what they did wrong. Whether it's a coworker or you in 6 months, they'll be grateful.

File: lib/booking_system.rb [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+require_relative 'reservation'
+require 'pry'
+
+module Hotel
+  class BookingSystem
+    attr_reader :reservations, :rooms
+
+    VALID_ROOM_NUMBERS = (1..20)
+    ROOM_COST = 200
+    DISCOUNT_ROOM_RATE = 150
```
Comment (droberts-sea) [inline lib/booking_system.rb:11]: Good use of constants for these.

File: lib/booking_system.rb [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+
+
+    ####
+    # room: an Integer for the room number
+    # check_in: a Date object for the check in day
+    # check_out: a Date object for the check out day
+    def book_room(room_number, check_in, check_out)
+      unless VALID_ROOM_NUMBERS.include?(room_number)
+        raise ArgumentError
+      end
```
Comment (droberts-sea) [inline lib/booking_system.rb:26]: Good use of vertical white space throughout this file - you do a good job of breaking up ideas, and it makes the whole thing much easier to read.

File: lib/booking_system.rb [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+      new_reservation = Hotel::Reservation.new(room_number, ROOM_COST, date_range)
+
+      @reservations << new_reservation
+      return new_reservation
+    end
+
+    def reservations_on_date(date)
+      reservations = []
+      @reservations.each do |reservation|
+        if reservation.within_reservation?(date)
```
Comment (droberts-sea) [inline lib/booking_system.rb:43]: This is a good example of loose coupling. Rather than having to pull out the start and end dates from the reservation and do a bunch of math, you call a method and let the `Reservation` take care of it. Good work.

File: lib/date_range.rb [hunk 5]
```diff
@@ -0,0 +1,24 @@
+require 'date'
+
+module Hotel
+  class DateRange
+    attr_reader :start_date, :end_date
+
+    def initialize(start_date, end_date)
+      if (end_date < start_date)
+        raise StandardError
+      end
```
Comment (droberts-sea) [inline lib/date_range.rb:10]: Instead of raising a `StandardError`, best practice is to define your own appropriately named exception type that inherits from `StandardError` and raise that. The reason is that `StandardError` is very broad, and if the code that calls this method has to `rescue StandardError`, that will catch things like `ArgumentError`s and `NameError`s that indicate true bugs.

File: lib/booking_system.rb [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+
+    def available_rooms(date_range)
+      # made a copy of the rooms array
+      rooms = @rooms.to_a
+
+      # iterating through date ranges, at each reservation, removing the room
+      # that is reserved by that reservation
+      (date_range.start_date...date_range.end_date).each do |date|
+        self.reservations_on_date(date).each do |reservation|
+          rooms.delete(reservation.room_number)
```
Comment (droberts-sea) [inline lib/booking_system.rb:58]: This works. However, there is a more efficient way to do it that doesn't involve looping through all the dates in the range. I challenge you to find it (and then go read the instructor implementation of Hotel to check your work).

File: spec/booking_system_spec.rb [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+  before do
+    @room_number = 12
+    @check_in = Date.new(2018, 8, 24)
+    @check_out = @check_in + 5
+    @booking_system = Hotel::BookingSystem.new
+  end
+
+  it "create a class for the booking system" do
+    @booking_system.must_be_kind_of Hotel::BookingSystem
+  end
```
Comment (droberts-sea) [inline spec/booking_system_spec.rb:14]: You should break these tests up a little. I typically have one big `describe` for the whole class, and then one sub-`describe` for each instance method.

File: spec/booking_system_spec.rb [hunk 8]
```diff
@@ -0,0 +1,58 @@
...
+    @booking_system.reservations.length.must_equal 1
+  end
+
+  it "returns all rooms" do
+    @booking_system.must_respond_to :rooms
+  end
+
+  it "returns list of reservations on a given date" do
+    @booking_system.reservations_on_date(@check_in).must_be_kind_of Array
+  end
```
Comment (droberts-sea) [inline spec/booking_system_spec.rb:41]: This doesn't check that you got the correct set of reservations back, just that you got an array. I would be interested in whether matching reservations were found, and whether non-matches were omitted. Another question: what happens if there are no matches?

File: spec/booking_system_spec.rb [hunk 9]
```diff
@@ -0,0 +1,58 @@
...
+
+  it "raises argument if room_number is outside range" do
+    expect{ @booking_system.book_room(50, @check_in, @check_out) }.must_raise ArgumentError
+  end
+
+  it "returns all reservations" do
+    @booking_system.book_room(@room_number, @check_in, @check_out)
+    @booking_system.must_respond_to :reservations
+    @booking_system.reservations.length.must_equal 1
```
Comment (droberts-sea) [inline spec/booking_system_spec.rb:31]: You might check that the length was 0 before making the reservation.

File: spec/booking_system_spec.rb [hunk 10]
```diff
@@ -0,0 +1,58 @@
...
+  it "create a class for the booking system" do
+    @booking_system.must_be_kind_of Hotel::BookingSystem
+  end
+
+  it "books a new room" do
+    new_reservation = @booking_system.book_room(@room_number, @check_in, @check_out)
+    new_reservation.must_be_kind_of Hotel::Reservation
+
+    @booking_system.reservations.length.must_equal 1
+  end
```
Comment (droberts-sea) [inline spec/booking_system_spec.rb:21]: I looks like you have stumbled on one of the secrets of minitest: wrapping values to assert on in `expect` isn't required.

However you should make a habit of doing so, for two reasons:
1. Most other testing frameworks (e.g. Jest for JavaScript, UnitTest for Python) do require something like an `expect`
1. Minitest is going to remove support for leaving `expect` out in the next major release

File: spec/booking_system_spec.rb [hunk 11]
```diff
@@ -0,0 +1,58 @@
...
+  it "returns a list of available rooms" do
+    date_range = Hotel::DateRange.new(@check_in, @check_out)
+    @booking_system.book_room(@room_number, @check_in, @check_out)
+    @booking_system.available_rooms(date_range).wont_include @room_number
+  end
+
+  it "raises ArgumentError if requested room is unavailable" do
+    @booking_system.book_room(@room_number, @check_in, @check_out)
+    expect { @booking_system.book_room(@room_number, @check_in, @check_out) }.must_raise ArgumentError
+  end
```
Comment (droberts-sea) [inline spec/booking_system_spec.rb:54]: For readability, you should break these exception expectations across multiple lines:

```ruby
expect {
  @booking_system.book_room(@room_number, @check_in, @check_out)
}.must_raise ArgumentError
```

File: spec/date_range_spec.rb [hunk 12]
```diff
@@ -0,0 +1,42 @@
...
+  end
+
+  it "raises a standard error if end date is before start date" do
+    expect{ Hotel::DateRange.new(@check_out, @check_in) }.must_raise StandardError
+  end
+
+  it "checks to see if a given date is within the date range" do
+
+    @date_range.include?(@check_in).must_equal true
+  end
```
Comment (droberts-sea) [inline spec/date_range_spec.rb:40]: I would like to see a more robust set of test cases here:

- If the date is way before, should return false
- If the date is one day before, should return false
- If the date is on the checkin day, should return true
- If the date is in the middle, should return true
- If the date is on the day before the checkout day, should return true
- If the date is on the checkout day, should return false
- If the date is after the checkout day, should return false

File: spec/reservation_spec.rb [hunk 13]
```diff
@@ -0,0 +1,41 @@
...
+  end
+
+  it "calculates total cost of stay" do
+    @reservation.stay_total_cost.must_equal 1000
+  end
+
+  it "checks to see if a given date is within a reservation" do
+
+    @reservation.within_reservation?(@check_in).must_equal true
+  end
```
Comment (droberts-sea) [inline spec/reservation_spec.rb:37]: Since this work is delegated to the reservation's date range, you don't need quite as robust of test coverage. However, I would still like to see at least one positive and negative case each.

**Comment by droberts-sea** [comment]
# Hotel
## What We're Looking For

Feature | Feedback
---     | ---
**Baseline** |
Used git regularly	| yes
Answer comprehension questions	| yes
**Design** |
Each class is responsible for a single piece of the program | yes
Classes are loosely coupled | yes
**Wave 1** |
List rooms | yes
Reserve a room for a given date range | yes
List reservations for a given date | yes
Calculate reservation price | yes
Invalid date range produces an error | yes
Test coverage | yes - missing some, see inline
**Wave 2** |
View available rooms for a given date range | yes
Reserving a room that is not available produces an error | yes
Test coverage | yes
**Wave 3** |
Create a block of rooms | no
Check if a block has rooms | no
Reserve a room from a block | no
Test coverage | no
**Fundamentals** | 
Names variables, classes and modules appropriately | yes
Understanding of variable scope - local vs instance | yes
Can create complex logical structures utilizing variables | yes
Appropriately uses methods to break down tasks into smaller simpler tasks | yes - good work!
Understands the differences between class and instance methods | yes
Appropriately uses iterators and enumerables | yes
Appropriately writes and utilizes classes | yes
Appropriately utilizes modules as a mechanism to encapsulate functionality | yes
**Wrap Up** |
There is a refactors.txt file | yes
The file provides a roadmap to future changes | yes
**Additional Feedback** | Good work overall! Your design is clean, your code is well-written, and it's clear the learning goals for this assignment were met. Your test coverage could be expanded somewhat, and it would have been nice to get wave 3 in, but in general I am happy with what I see. Keep up the hard work!
